### PR TITLE
Add __CXXLIB__ special token

### PR DESCRIPTION
**Pull request by TurkeyMan**

### CODE REVIEW

**Comment by TurkeyMan** [comment]
Alternate solution for: https://github.com/dlang/dmd/pull/8655

**Review by jacob-carlborg — CHANGES_REQUESTED** [review]
There are no tests.

**Comment by TurkeyMan** [comment]
I don't know how to test this?

**Comment by jacob-carlborg** [comment]
> I don't know how to test this?

Use `REQUIRED_ARGS` to specify a value for `mscrtlib` and assert that `__MSCRT_VER__` is what's passed to `mscrtlib`. Something like:

```d
// REQUIRED_ARGS: -mscrtlib=foo
static assert(__MSCRT_VER__ == "foo");
```

Would that work? Not sure how you should limit the test to Windows only though.

**Comment by TurkeyMan** [comment]
That will only work on windows+coff... the test should not pass any args for !windows || omf.

**Comment by jacob-carlborg** [comment]
I see that you found a solution for the platform specific testing, great 👍.

File: test/compilable/cpp.d [hunk 1]
```diff
@@ -0,0 +1,19 @@
+
+// Test C++ features.
+
+version(Windows)
+{
+    version(D_LP64)
```
Comment (rainers) [inline test/compilable/cpp.d]: You can just use CRuntime_Microsoft instead.

File: test/compilable/cpp.d [hunk 2]
```diff
@@ -0,0 +1,19 @@
+
+// Test C++ features.
```
Comment (jacob-carlborg) [inline test/compilable/cpp.d]: Is this a C++ feature?

Comment (TurkeyMan) [reply test/compilable/cpp.d]: Kinda.

Comment (TurkeyMan) [reply test/compilable/cpp.d]: Welcome to suggest another location?

File: test/compilable/cpp.d (hunk 1 above)
Comment (TurkeyMan) [reply test/compilable/cpp.d]: Oh yeah... we have those now huh ;)

**Comment by jacob-carlborg** [comment]
Should have a corresponding PR for the spec.

**Comment by TurkeyMan** [comment]
https://github.com/dlang/dlang.org/pull/2469

**Comment by TurkeyMan** [comment]
Is this good? Not sure why CI never works... doesn't seem to be my fault.

**Review by WalterBright — CHANGES_REQUESTED** [review]
I'll approve it with changing `__MSCRTLIB__` to `__CRTLIB__` so it'll work with any future  CRT library.

**Comment by TurkeyMan** [comment]
Incidentally, this is actually about the C++RT... `msvcrt` and friends are the C++RT, the CRT is called `ucrt`.
Should I make this a pair for the C and C++ runtimes? Or should I just call this `__CXXRTLIB__`?
These are kinda parallel to all those versions that we have for detecting various C/C++ runtimes; should these strings attempt to be populated in all cases?

**Comment by WalterBright** [comment]
I don't know about other systems, but for Digital Mars C/C++ the crt C and C++ is combined into snn.lib. But I suppose they can be separate, so I'd go with __CRTLIB__ and __CXXRTLIB__.

**Comment by WalterBright** [comment]
I see __CXXLIB__ but not __CLIB__ ?

**Comment by TurkeyMan** [comment]
I've added that guy in the follow-up patch, since it can't meaningfully have any effect until additional plumbing carries those values from the system specification code through into where they're used.
Seemed like no point adding it until it does something.

**Comment by TurkeyMan** [comment]
just working it through now, and i need to confirm the names for all the DMD targets to the best of my ability.

**Comment by TurkeyMan** [comment]
I'm having second thoughts about the CLIB one actually, because unlike the C++ lib which is hard-wired into toolchains, and C lib is ABI agnostic, and potentially selected only at link-time, so the compiler may not have knowledge of which C lib is used up-front...
C++ lib is not toolchain agnostic, and somewhat defines the ABI, so we need to know (and can know) at compile time which is selected.

**Comment by WalterBright** [comment]
Well, make up your mind :-)

**Comment by TurkeyMan** [comment]
;) ... well I just liked the symmetry of the thing... but it's not actually correct.

I think this as is for now. We can extend to other platforms if/when it becomes useful.
If we extend to all targets eagerly, we're creating a lot of work for the GDC/LDC guys trying to specify them for all possible targets, which are extremely numerous! Let's wait and see if we need that for other platforms too.

**Comment by WalterBright** [comment]
I agree, I just wanted the name change because there was no reason to make it Microsoft specific.

**Comment by WalterBright** [comment]
I also like this a lot better than the proliferation of versions, which made my stomach churn with "there's gotta be a better way".

**Comment by TurkeyMan** [comment]
Well, as far as I know, it _is_ actually Microsoft specific... I don't know of any other instances where this knowledge is required to compile against the correct ABI.

I agree too, this is much better than the versions.

And I think this is the last of my supporting trivia; I can get back on with it now ;)

**Review by WalterBright — APPROVED** [review]

**Comment by TurkeyMan** [comment]
Danke!

**Comment by WalterBright** [comment]
Bitte

**Comment by ibuclaw** [comment]
> so the compiler may not have knowledge of which C lib is used up-front

Yes it does, as each library has its own macros to identify them.

Is `__CLIB__` referring to the C runtime library or the C standard library? The runtime library is compiler specific (libgcc) but the standard library is interchangeable (glibc, musl).

File: src/dmd/lexer.d [hunk 3]
```diff
@@ -530,6 +530,24 @@ class Lexer : ErrorHandler
                             t.ustring = global.compiler.vendor;
                             goto Lstr;
                         }
+                        else if (id == Id.CXXLIB)
+                        {
+                            // TODO: other targets might also like to populate this with knowledge for library authors
+                            //       additional plumbing required...
```
[the remark this answers is not in the document]
Comment (TurkeyMan) [reply src/dmd/lexer.d]: I followed the example from the other similar identifiers... it was very low-impact. What's the advantage of doing it elsewhere?

**Comment by TurkeyMan** [comment]
It's hard to know what's useful... probably neither is useful.
Are there cases where they might change and library code needs to know about it?

**Comment by TurkeyMan** [comment]
I did what I know is needed, and I didn't want to over-do it, because that would just cascade work onto you and the LDC folks, and I'm not sure there's any use to that effort.

**Comment by jacob-carlborg** [comment]
We already have version identifiers for the different C standard libraries we're supporting. Would `__CLIB__` be something else?

File: src/dmd/lexer.d (hunk 3 above)
[the remark this answers is not in the document]
Comment (ibuclaw) [reply src/dmd/lexer.d]: There should be nothing target-specific in the front-end, period.

[the remark this answers is not in the document]
Comment (TurkeyMan) [reply src/dmd/lexer.d]: Okay. I mean, that certainly sounds good... but there is heaps of target-specific stuff in there.
I was just trying to write the lowest-impact patch. Nobody gave me guidance on how to do it, I just followed precedent... which is turns out seems to be a bad idea perhaps 100% of the time in my experience so far ;)

[the remark this answers is not in the document]
Comment (jacob-carlborg) [reply src/dmd/lexer.d]: Yeah, we’re in the process of trying to improve the code base. Unfortunately you’re caught in the middle of this.

[the remark this answers is not in the document]
Comment (ibuclaw) [reply src/dmd/lexer.d]: Then stop following precedence then. :-) 

At some point in the future I will remove all `global.params.isXXX` code, as when making decisions in the language, asking "am I generating code for Windows?" is the wrong question. Correct question would be: "What is the C++ library I'm linking to?"

The answer to which can be implemented in Target and be as dmd-centric as you like, as each compiler backend should be providing their own implementation of that module.

[the remark this answers is not in the document]
Comment (ibuclaw) [reply src/dmd/lexer.d]: I still think though that the parser and lexer should have a blanket ban on all things target related to make it as agnostic as possible.  Forwarding the job of handling the decisions to the semantic passes.

Hence, adding a new placeholder AST type to wrap this information.

[the remark this answers is not in the document]
Comment (jacob-carlborg) [reply src/dmd/lexer.d]: We could also have a generic place for this, we could use `__traits` or come up with a similar system. With `__traits` I like that we only reserve one name (`__traits`) and the rest of the symbols are scoped within `__traits`, compared to this approach of using global symbols.

[the remark this answers is not in the document]
Comment (ibuclaw) [reply src/dmd/lexer.d]: Yeah. @TurkeyMan - what do you have against __traits?

[the remark this answers is not in the document]
Comment (TurkeyMan) [reply src/dmd/lexer.d]: I have nothing against `__traits`, I don't care how this is implemented... I just did it how people suggested to do it. In this case, I rewrite my original patch to use `__IDENTIFIER__` on specific request...

**Pull request merged**